### PR TITLE
make chaos dashboard be compliable with namespace scoped mode

### PR DESCRIPTION
**Pull request by STRRL**
### What problem does this PR solve?

Fix  https://github.com/chaos-mesh/chaos-mesh/issues/1107, chaos-dashboard is uncompliable  with namespace scoped mode.

### What is changed and how does it work?
- Add ENV to chaos-dashboard in helm deployment yaml.
- Adding new options while creating `Manager`, only watching ONE specific namespace.
- Change some `List` action while user query something from k8s api(detail in Side effects).
> appending at 2020-10-29
- change `/api/common/namespace` to `/api/common/chaos-available-namespaces`, means only return namespaces which could inject chaos (chaos experiment explosion scope)
- remove permission about `list` on `namespaces` with namespace scoped mode

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

It changes `/api/experiments` behavior under namespace scoped mode: origin parameter `namespace` will be replaced by `targetNamespace`.

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: helm/chaos-mesh/templates/chaos-dashboard-deployment.yaml [hunk 1]
```diff
@@ -48,6 +48,18 @@ spec:
...
+            - name: CLUSTER_SCOPED
+              value: "{{ .Values.clusterScoped }}"
+            {{- if .Values.controllerManager.allowedNamespaces }}
+            - name: ALLOWED_NAMESPACES
+              value: {{ .Values.controllerManager.allowedNamespaces }}
+            {{- end }}
+            {{- if .Values.controllerManager.ignoredNamespaces }}
+            - name: IGNORED_NAMESPACES
+              value: {{ .Values.controllerManager.ignoredNamespaces }}
+            {{- end }}
```
Comment (WangXiangUSTC) [inline helm/chaos-mesh/templates/chaos-dashboard-deployment.yaml:62]: seems that `ALLOWED_NAMESPACES` and `IGNORED_NAMESPACES` is useless

**Review by WangXiangUSTC — COMMENTED** [review]
rest LGTM

File: helm/chaos-mesh/templates/chaos-dashboard-deployment.yaml (hunk 1 above)
Comment (STRRL) [reply helm/chaos-mesh/templates/chaos-dashboard-deployment.yaml:62]: These ENV will  inject into `common.ControllerCfg` which was dependent in `pkg/utils/selector.go`.

Many API in `pkg/apiserver/common/common.go` use `pkgutils.SelectPods` which located in `pkg/utils/selector.go`. For keeping same action, I think we should set these ENVs.

This is a bad-smell  which already pointed by https://github.com/chaos-mesh/chaos-mesh/issues/1072 😥.

**Review by WangXiangUSTC — DISMISSED** [review]
LGTM

**Comment by STRRL** [comment]
PTAL @cwen0

**Comment by cwen0** [comment]
@STRRL  Does this configuration affect the [common API](https://github.com/chaos-mesh/chaos-mesh/blob/master/pkg/apiserver/common/common.go)?

**Comment by STRRL** [comment]
> @STRRL Does this configuration affect the [common API](https://github.com/chaos-mesh/chaos-mesh/blob/master/pkg/apiserver/common/common.go)?

Yes, kind of fixing a bug.

There are two ENV `ALLOWED_NAMESPACES` and `IGNORED_NAMESPACES ` are NOT set in old deployment yaml file. These ENVs are needed by `pkgutils.SelectPods()` in common API.

**Comment by cwen0** [comment]
> > @STRRL Does this configuration affect the [common API](https://github.com/chaos-mesh/chaos-mesh/blob/master/pkg/apiserver/common/common.go)?
> 
> Yes, kind of fixing a bug.
> 
> There are two ENV `ALLOWED_NAMESPACES` and `IGNORED_NAMESPACES ` are NOT set in old deployment yaml file. These ENVs are needed by `pkgutils.SelectPods()` in common API.

If we set namespace scope, `common.getNamespaces` may return a error. Can we avoid this behavior?

**Comment by STRRL** [comment]
> If we set namespace scope, `common.getNamespaces` may return a error. Can we avoid this behavior?

Currently service account dose hold the permission about `list` with `namespace` in cluster level, [link](https://github.com/chaos-mesh/chaos-mesh/blob/4cde3cafcee1158752176f5b035389edb9f2e6cb/helm/chaos-mesh/templates/controller-manager-rbac.yaml#L46). So I don't think it will throw an error.

**Comment by STRRL** [comment]
More changes:
- change `/api/common/namespace` to `/api/common/chaos-available-namespaces`, means only return namespaces which could inject chaos (chaos experiment explosion scope)
- remove permission about `list` on `namespaces` with namespace scoped mode

PTAL @cwen0  @WangXiangUSTC @g1eny0ung

File: pkg/config/dashboard.go [hunk 2]
```diff
@@ -29,6 +29,11 @@ type ChaosDashboardConfig struct {
 	EnableLeaderElection bool   `envconfig:"ENABLE_LEADER_ELECTION"`
 	Database             *DatabaseConfig
 	PersistTTL           *PersistTTLConfig
+	// ClusterScoped means control Chaos Object in cluster level(all namespace),
+	ClusterScoped bool `envconfig:"CLUSTER_SCOPED" default:"true"`
+	// TargetNamespace is the target namespace to injecting chaos.
+	// It only works with ClusterScoped is false;
+	TargetNamespace string `envconfig:"TARGET_NAMESPACE" default:""`
```
Comment (g1eny0ung) [inline pkg/config/dashboard.go:36]: Should we set the `TARGET_NAMESPACE` to `chaos-testing`? I noticed that the installation script defaults to chaos-testing.

File: install.sh [hunk 3]
```diff
@@ -951,7 +951,6 @@ metadata:
 rules:
   - apiGroups: [ "" ]
     resources:
-      - namespaces
```
Comment (WangXiangUSTC) [inline install.sh:954]: when `CLUSTER_SCOPED` is true, still need the privileged to list namespace

File: pkg/config/dashboard.go (hunk 2 above)
Comment (STRRL) [reply pkg/config/dashboard.go:36]: It's configured by `controllerManager.targetNamespace` in `helm/chaos-mesh/values.yaml`, [link](https://github.com/chaos-mesh/chaos-mesh/blob/master/helm/chaos-mesh/values.yaml#L43).

We have a bash script that rendering `instasll.sh` from helm files.

File: install.sh (hunk 3 above)
Comment (STRRL) [reply install.sh:954]: I will fix it. 🙈

File: pkg/config/dashboard.go (hunk 2 above)
Comment (g1eny0ung) [reply pkg/config/dashboard.go:36]: Understood, thx for the explanation.~

File: pkg/apiserver/common/common.go [hunk 4]
```diff
@@ -65,7 +65,7 @@ func Register(r *gin.RouterGroup, s *Service) {
 	endpoint := r.Group("/common")
 
 	endpoint.POST("/pods", s.listPods)
-	endpoint.GET("/namespaces", s.getNamespaces)
+	endpoint.GET("/chaos-available-namespaces", s.getChaosAvailableNamespaces)
```
Comment (WangXiangUSTC) [inline pkg/apiserver/common/common.go:68]: how about keep the `/namespaces` and add the new API `/chaos-available-namespaces`

Comment (STRRL) [reply pkg/apiserver/common/common.go:68]: Great! I am working on this.

Comment (STRRL) [reply pkg/apiserver/common/common.go:69]: Hi @WangXiangUSTC, it finished in 4a056c7 , please take a look.

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

**Comment by WangXiangUSTC** [comment]
Does this PR need cherry pick to 1.0?

**Comment by STRRL** [comment]
> Does this PR need cherry pick to 1.0?

Yes it is.

**Review by g1eny0ung — APPROVED** [review]
LGTM

**Pull request merged**